internal/replication/blockreplicator.go

```diff
@@ -290,6 +306,89 @@ Event_Loop:
 	br.lg.Info("Exiting block replicator event loop")
 }
 
+// When a node lags behind the cluster more than the last checkpoint of the leader, the leader will send a snapshot to
+// it. A snapshot is a block with some raft information. A received snapshot serves as a trigger for the node to
+// perform catch-up, or state transfer. It will contact one of the active members of the cluster (preferably the
+// leader), and will request the missing blocks up to the block indicated by the snapshot.
+func (br *BlockReplicator) catchUp(snap raftpb.Snapshot) error {
+	if snap.Metadata.Index <= br.appliedIndex {
+		br.lg.Debugf("Skip snapshot taken at index %d, because it is behind current applied index %d", snap.Metadata.Index, br.appliedIndex)
+		return nil
+	}
+
+	var snapBlock = &types.Block{}
+	if err := proto.Unmarshal(snap.Data, snapBlock); err != nil {
+		return errors.Errorf("failed to unmarshal snapshot data to block: %s", err)
+	}
+
+	br.lg.Debugf("last block: %+v", snapBlock)
+	br.lg.Debugf("snap block: %+v", snapBlock)
+	if br.lastBlock.Header.BaseHeader.Number >= snapBlock.Header.BaseHeader.Number {
+		br.lg.Errorf("Snapshot is at block [%d], local block number is %d, no catch-up needed", snapBlock.Header.BaseHeader.Number, br.lastBlock.Header.BaseHeader.Number)
+		return nil
+	}
+
+	br.lg.Infof("Starting state transfer; From block: %d, index %d; To block: %d, index: %d",
+		br.lastBlock.Header.BaseHeader.Number, br.appliedIndex, snapBlock.Header.BaseHeader.Number, snap.Metadata.Index)
+	br.confState = snap.Metadata.ConfState
+	br.appliedIndex = snap.Metadata.Index
+
+	for nextBlockNumber := br.lastBlock.Header.BaseHeader.Number + 1; nextBlockNumber <= snapBlock.Header.BaseHeader.Number; {
+		var blocks []*types.Block
+		var err error
+		blocksReadyCh := make(chan struct{})
+		ctx, cancel := context.WithCancel(context.Background())
+
+		//Try to pull some blocks in a go-routine so that we may cancel it if the server shuts down.
+		go func() {
+			defer close(blocksReadyCh)
+			blocks, err = br.transport.PullBlocks(ctx, nextBlockNumber, snapBlock.Header.BaseHeader.Number, br.GetLeaderID())
```

Comment:
I will document better.

Why from the leader first?
- simplicity, as we know that the leader is on line as it sent us the snapshot
- applying pressure on the leader slows down the cluster and reduces the chance that the cluster will run away from the node trying to catchup, and it will never be able to catch up. (this is what Raft does, in a sense). This is important, since as long as member does not catch up the cluster is in a state of reduced fault tolerance.
- we have in our road map a plan to implement "observers", they will be able to synch with the cluster by using followers rather than the leader, if we so chose.